src/logic/shift.rs

```diff
@@ -0,0 +1,886 @@
+
+    /// Arithmetically right-shifts this `ApInt` by the given `shift_amount` bits.
+    /// 
+    /// # Note
+    /// 
+    /// Arithmetic right shifts copy the sign bit to the most significant bits, and thus can act as
+    /// a floored division by a power of two for the **signed** interpretation of `ApInt`s.
+    /// 
+    /// # Errors
+    /// 
+    /// - If the given `shift_amount` is invalid for the bit width of this `ApInt`.
+    pub fn wrapping_ashr_assign<S>(&mut self, shift_amount: S) -> Result<()>
+        where S: Into<ShiftAmount>
+    {
+        if self.sign_bit() == false {
+            return self.wrapping_lshr_assign(shift_amount)
+        }
+        let shift_amount = shift_amount.into();
+        shift_amount.verify_shift_amount(self)?;
+        let width = self.width();
+        match self.access_data_mut() {
+            DataAccessMut::Inl(digit) => {
+                let mut signed = digit.clone();
+                signed.sign_extend_from(width).unwrap();
+                let signed = signed.repr() as i64;
+                let shifted = signed >> shift_amount.to_usize();
+                *digit.repr_mut() = shifted as u64;
+            }
+            DataAccessMut::Ext(digits) => {
+                let digit_steps = shift_amount.digit_steps();
+                if digit_steps != 0 {
+                    digits.rotate_left(digit_steps);
+                    digits.iter_mut()
+                          .rev()
+                          .take(digit_steps)
+                          .for_each(|d| *d = Digit::all_set());
+                }
+                let bit_steps = shift_amount.bit_steps();
+                if bit_steps > 0 {
+                    let mut borrow = 0xFFFF_FFFF_FFFF_FFFF << (Digit::BITS - bit_steps);
+                    for elem in digits.iter_mut().rev().skip(digit_steps) {
+                        let repr = elem.repr();
+                        let new_borrow = repr << (Digit::BITS - bit_steps);
+                        *elem = Digit((repr >> bit_steps) | borrow);
+                        borrow = new_borrow;
+                    }
+                }
+            }
+        }
+        self.clear_unused_bits();
+        Ok(())
+    }
+
+    /// Arithmetically right-shifts this `ApInt` by the given `shift_amount` bits
+    /// and returns the result.
+    /// 
+    /// # Note
+    /// 
+    /// Arithmetic right shifts copy the sign bit to the most significant bits, and thus can act as
+    /// a floored division by a power of two for the **signed** interpretation of `ApInt`s.
+    /// 
+    /// # Errors
+    /// 
+    /// - If the given `shift_amount` is invalid for the bit width of this `ApInt`.
+    pub fn into_wrapping_ashr<S>(self, shift_amount: S) -> Result<ApInt>
+        where S: Into<ShiftAmount>
+    {
+        try_forward_bin_mut_impl(self, shift_amount, ApInt::wrapping_ashr_assign)
+    }
+
+    /// Circularly left-rotates this `ApInt` by the given `shift_amount` bits. In other words, the
```

Comment:
Could you provide some example codes or example scenarios where one could really use this functionality? Since Rust integers do not have support for this I am curious.

Reply:
Rust integers do have support for this, e.g. `i64::rotate_left`. Rotate shifts are sometimes used in places where bit vectors are being XORed to themselves, e.g. in certain random number generators.

Reply:
Okay nice, didn't know that!